### PR TITLE
vmap: temporarily disable support for random functions

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42617 vmap: temporarily disable support for random functions**

While we figure out the random plan, I want to initially disable
support for random operations. This is because there is an ambiguity in
what randomness means. For example,

```
tensor = torch.zeros(B0, 1)
vmap(lambda t: t.normal_())(tensor)
```

in the above example, should tensor[0] and tensor[1] be equal (i.e.,
use the same random seed), or should they be different?

The mechanism for disabling random support is as follows:
- We add a new dispatch key called VmapMode
- Whenever we're inside vmap, we enable VmapMode for all tensors.
This is done via at::VmapMode::increment_nesting and
at::VmapMode::decrement_nesting.
- DispatchKey::VmapMode's fallback kernel is the fallthrough kernel.
- We register kernels that raise errors for all random functions on
DispatchKey::VmapMode. This way, whenever someone calls a random
function on any tensor (not just BatchedTensors) inside of a vmap block,
an error gets thrown.

Test Plan:
- pytest test/test_vmap.py -v -k "Operators"

Differential Revision: [D22954840](https://our.internmc.facebook.com/intern/diff/D22954840)

### CODE REVIEW

File: aten/src/ATen/VmapModeRegistrations.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+TORCH_LIBRARY_IMPL(_, VmapMode, m) {
+  m.fallback(torch::CppFunction::makeFallthrough());
+}
+
+TORCH_LIBRARY_IMPL(aten, VmapMode, m) {
+  // NB: I'd really like to register a special kernel like
+  // CppFunction::makeNamedNotSupported() to avoid listing out the types of everything.
+  // However, registering e.g. CppFunction::makeNamedNotSupported() as an implementation
+  // only works for operators that support boxing.
+#define TENSOROPTIONS c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>
```
Comment (ezyang) [inline aten/src/ATen/VmapModeRegistrations.cpp:42]: cc @smessmer @bhosmer @gchanan Downside of exploded tensor options

File: aten/src/ATen/VmapMode.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+namespace at {
+
+// VmapMode contains a thread local count of how many nested vmaps
+// we are currently inside. That number is known as the `vmap level`.
+// VmapMode is used in the implementation of the Python `torch.vmap` API.
+//
+// NOTE: this is NOT the c++ api for torch.vmap. That doesn't exist yet.
```
Comment (ezyang) [inline aten/src/ATen/VmapMode.h:11]: If this isn't intended to be public API, put it in a private namespace like `impl`

File: aten/src/ATen/VmapModeRegistrations.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+TORCH_LIBRARY_IMPL(_, VmapMode, m) {
+  m.fallback(torch::CppFunction::makeFallthrough());
+}
+
+TORCH_LIBRARY_IMPL(aten, VmapMode, m) {
+  // NB: I'd really like to register a special kernel like
+  // CppFunction::makeNamedNotSupported() to avoid listing out the types of everything.
+  // However, registering e.g. CppFunction::makeNamedNotSupported() as an implementation
+  // only works for operators that support boxing.
```
Comment (ezyang) [inline aten/src/ATen/VmapModeRegistrations.cpp:41]: This seems like an oversight; there's no reason this shouldn't work. File us a bug?

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/VmapModeRegistrations.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/VmapModeRegistrations.cpp:42]: I really wanted to do an e.g. `decltype(native::randn)` here to avoid listing out the types. However there were two problems with that:
1) we've got a lot of overloads
2) native::randn takes `const TensorOptions&`, so a decltype wouldn't give the correct thing

File: aten/src/ATen/VmapModeRegistrations.cpp (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/VmapModeRegistrations.cpp:41]: Bug filed to #42818
